feat: scroll restoration (#1)
moxystudio/next-scroll-behavior: README.md

## Installation

```sh
$ npm install @moxy/next-scroll-restoration
```

This library is written in modern JavaScript and is published in both CommonJS and ES module transpiled variants. If you target older browsers please make sure to transpile accordingly.

## Motivation

There are some cases where you need to take control on how your application scroll is handled; namely, you may want to restore scroll when user is navigating within your application pages, but you need to do extra work before or after the page has changed, either by using some sort of page transition or any other feature.

`next-scroll-restoration` provides a built-in custom scroll behavior that aims to make it easy to restore the user scroll position based on location history.

This package is built on top of [scroll-behavior](https://www.npmjs.com/package/scroll-behavior) and it's meant to be used in `Nextjs` applications.

## Usage

You can simply import the package and update the scroll whenever your need.

```js
// ...

import { getScrollBehavior } from '@moxy/next-scroll-behavior';

// ...

> const ScrollBehavior = getScrollBehavior();

Reviewer: You forgot to change to `scrollBehavior`. Please check this on other places.

const updateScrollAfterExtraWork = useCallBack(() => {
    // ...
    
    ScrollBehavior.updateScroll();
    
    // ...

> },[])

Reviewer: ```suggestion
},[]);
```

```

This library is written in modern JavaScript and is published in both CommonJS and ES module transpiled variants. If you target older browsers please make sure to transpile accordingly.

## Motivation

There are some cases where you need to take control on how your application scroll is handled; namely, you may want to restore scroll when user is navigating within your application pages, but you need to do extra work before or after the page has changed, either by using some sort of page transition or any other feature.

`next-scroll-restoration` provides a built-in custom scroll behavior that aims to make it easy to restore the user scroll position based on location history.

This package is built on top of [scroll-behavior](https://www.npmjs.com/package/scroll-behavior) and it's meant to be used in `Nextjs` applications.

## Usage

You can simply import the package and update the scroll whenever your need.

```js
// ...

import { getScrollBehavior } from '@moxy/next-scroll-behavior';

// ...

const ScrollBehavior = getScrollBehavior();

const updateScrollAfterExtraWork = useCallBack(() => {
    // ...
    
    ScrollBehavior.updateScroll();
    
    // ...
},[])
```

> > ⚠️ This package monkey patches `Next Link` to ensure that the default scroll top behavior is disabled. You can enable `Link` default behavior by passing `disableNextLinkScroll` as `true` in the option argument when calling `getScrollBehavior`.

Reviewer: ```suggestion
> ⚠️ This package monkey patches Next.js `<Link />` component, changing the `scroll` prop default value to `false`. You can disable this behavior by passing `disableNextLinkScroll` as `true` in the option argument when calling `getScrollBehavior()`.
```

>
> Bear in mind that the package also sets `history.scrollRestoration` to `manual`. 

## How it works

This package works by actively listening to `Nextjs` router events on `routeChangeStart`, writing the scroll values associated with the current `location` in the `Session Storage` and reading these values whenever `updateScroll` is called.

## API

The function accepts an `options` object argument.

> | Option Key | Default | Description |
> | --- | ---- | ----------- |

Reviewer: ```suggestion
| Option Key | Default | Description |
| ---------- | ------- | ----------- |
```

> | `disableNextLinkScroll` | `true` | Control `Next Link` default scroll behavior |

Reviewer: ```suggestion
| `disableNextLinkScroll` | `true` | Enables or disables changing the default `scroll` prop of Next.js `<Link />` component |
```

```js

const scrollBehaviorOptions = {
    disableNextLinkScroll: false;
}

const ScrollBehavior = getScrollBehavior(scrollBehaviorOptions);

> 

Reviewer: ```suggestion
```

```js
// ...

import getScrollBehavior from '@moxy/next-scroll-behavior';

// ...

useEffect(() => {
    scrollBehaviorRef.current = getScrollBehavior();

    return () => {
        scrollBehaviorRef.current.stop();
    };
}, []);

// ...

const updateScrollAfterExtraWork = useCallBack(() => {
    // ...
    
    scrollBehaviorRef.current.updateScroll();
    
    // ...
},[]);
```

> ⚠️ This package monkey patches Next.js `<Link />` component, changing the `scroll` prop default value to `false`. You can disable this behavior by passing `disableNextLinkScroll` as `true` in the option argument when calling `getScrollBehavior()`.
>
> Bear in mind that the package also sets `history.scrollRestoration` to `manual`. 

## How it works

This package works by actively listening to `Nextjs` router events on `routeChangeStart`, writing the scroll values associated with the current `location` in the `Session Storage` and reading these values whenever `updateScroll` is called.

## API

The function accepts an `options` object argument.

| Option Key | Default | Description |
| ---------- | ------- | ----------- |

> | `disableNextLinkScroll` | `true` | `disableNextLinkScroll` | `true` | Enables or disables changing the default `scroll` prop of Next.js `<Link />` component |a

Reviewer: ```suggestion
| `disableNextLinkScroll` | `true` | `disableNextLinkScroll` | `true` | Enables or disables changing the default `scroll` prop of Next.js `<Link />` component |
```

[npm-url]:https://npmjs.org/package/@moxy/next-scroll-restoration
[downloads-image]:https://img.shields.io/npm/dm/@moxy/next-scroll-restoration.svg
[npm-image]:https://img.shields.io/npm/v/@moxy/next-scroll-restoration.svg
[build-status-url]:https://github.com/moxystudio/next-scroll-restoration/actions
[build-status-image]:https://img.shields.io/github/workflow/status/moxystudio/next-scroll-restoration/Node%20CI/master
[codecov-url]:https://codecov.io/gh/moxystudio/next-scroll-restoration
[codecov-image]:https://img.shields.io/codecov/c/github/moxystudio/next-scroll-restoration/master.svg
[david-dm-url]:https://david-dm.org/moxystudio/next-scroll-restoration
[david-dm-image]:https://img.shields.io/david/moxystudio/next-scroll-restoration.svg
[david-dm-dev-url]:https://david-dm.org/moxystudio/next-scroll-restoration?type=dev
[david-dm-dev-image]:https://img.shields.io/david/dev/moxystudio/next-scroll-restoration.svg

> `next-scroll-restoration`provides a built in custom scroll behavior that aims to make it easy to restore the user scroll position based on location history.

Reviewer 2: ```suggestion
`next-scroll-restoration` provides a built in custom scroll behavior that aims to make it easy to restore the user scroll position based on location history.
```

Author: ✅

## Installation

```sh
$ npm install @moxy/next-scroll-restoration
```

This library is written in modern JavaScript and is published in both CommonJS and ES module transpiled variants. If you target older browsers please make sure to transpile accordingly.

## Motivation

> There are some case where you need to take control on how your application scroll is handled, namely you want to restore scroll when user is navigating within your application pages but you need to do extra work before or after the page has changed either by using some sort of page transition or any other feature.

Reviewer 2: ```suggestion
There are some cases where you need to take control on how your application scroll is handled; namely, you may want to restore scroll when user is navigating within your application pages, but you need to do extra work before or after the page has changed, either by using some sort of page transition or any other feature.
```

> `next-scroll-restoration`provides a built in custom scroll behavior that aims to make it easy to restore the user scroll position based on location history.

Reviewer 2: ```suggestion
`next-scroll-restoration` provides a built-in custom scroll behavior that aims to make it easy to restore the user scroll position based on location history.
```

Author: ✅

> This package is built on top of [scroll-behavior](https://www.npmjs.com/package/scroll-behavior) and it's meant to be used in `Nextjs`applications.

Reviewer 2: ```suggestion
This package is built on top of [scroll-behavior](https://www.npmjs.com/package/scroll-behavior) and it's meant to be used in `Nextjs` applications.
```

Author: ✅

## Usage

> Firstly ensure that your `history.scrollRestoration` is set as `manual`within your main App file.

Reviewer 2: ```suggestion
First ensure that your `history.scrollRestoration` is set as `manual` in your main App file.
```

Author: ✅

```js
if (typeof history !== 'undefined') {
    history.scrollRestoration = 'manual';
}
```

Then you can simply import the package and update the scroll whenever your need.

```js
// ...

import { getScrollBehavior } from '@moxy/next-scroll-behavior';

// ...

const ScrollBehavior = getScrollBehavior();

const updateScrollAfterExtraWork = useCallBack(() => {
    // ...
    
    ScrollBehavior.updateScroll();
    
    // ...
},[])
```

## How it works

> This package works by actively listening to `Nextjs` router events on`routeChangeStart` and unlistening whenever`routeChangeStart`events off, writing the scroll values associated with the current `location` in the`Session Storage` and reading this values whenever `updateScroll` is called.

Reviewer 2: ```suggestion
This package works by actively listening to `Nextjs` router events on `routeChangeStart` and unlistening whenever `routeChangeStart` events off, writing the scroll values associated with the current `location` in the `Session Storage` and reading these values whenever `updateScroll` is called.
```

Reviewer 2: Additionally, I think you may have omitted some words from this part:

```
and unlistening whenever `routeChangeStart`events off, [...]
```

I don't really understand what you're trying to convey here.

Author: Updated ✅ that sentence was not needed at all.
